gateway-ha/src/main/java/io/trino/gateway/ha/router/TrinoRequestUser.java

```diff
@@ -0,0 +1,192 @@
+        if (header != null) {
+            return Optional.of(header);
+        }
+
+        Optional<String> user = extractUserFromAuthorizationHeader(request.getHeader("Authorization"), userField);
+        if (user.isPresent()) {
+            return user;
+        }
+
+        return extractUserFromCookies(request, userField);
+    }
+
+    private Optional<String> extractUserFromAuthorizationHeader(String header, String userField)
+    {
+        if (header == null) {
+            return Optional.empty();
+        }
+
+        if (header.contains("Basic")) {
+            return Optional.of(new String(Base64.getDecoder().decode(header.split(" ")[1]), StandardCharsets.UTF_8).split(":")[0]);
+        }
+
+        if (header.toLowerCase().contains("bearer")) {
+            return extractUserFromBearerAuth(header, userField);
+        }
+        return Optional.empty();
+    }
+
+    private Optional<String> extractUserFromBearerAuth(String header, String userField)
+    {
+        log.debug("Trying to extract user from bearer token");
+        int space = header.indexOf(' ');
+        if ((space < 0) || !header.substring(0, space).equalsIgnoreCase("bearer")) {
+            return Optional.empty();
+        }
+
+        String token = header.substring(space + 1).trim();
+        ObjectMapper mapper = new ObjectMapper();
+
+        if (header.split("\\.").length == 3) { //this is probably a JWS
+            log.debug("Trying to extract from JWS");
+            token = header.split("\\.")[1];
+            try {
+                JsonNode node = mapper.readTree(new String(Base64.getDecoder().decode(token)));
+                if (node.has(userField)) {
+                    log.debug("Trying to extract user from JWS json. User: %s", node.get(userField).asText());
+                    return Optional.of(node.get(userField).asText());
+                }
+            }
+            catch (JsonProcessingException e) {
+                log.warn("Could not deserialize bearer token as json");
+            }
+        }
+
+        if (isGetUserFromOAuth) {
+            String responseBody = getUserInfo(token);
+            if (responseBody.contains(userField)) {
+                mapper = new ObjectMapper();
+                try {
+                    JsonNode node = mapper.readTree(responseBody);
+                    if (node.has(userField)) {
+                        log.debug("Trying to extract user from json. User: %s", node.get(userField).asText());
+                        return Optional.of(node.get(userField).asText());
+                    }
+                }
+                catch (JsonProcessingException ex) {
+                    log.debug("Could not deserialize token info response to json: %s", responseBody);
+                }
+            }
+        }
+
+        return Optional.empty();
+    }
+
+    private String getUserInfo(String token)
+    {
+        OkHttpClient client = new OkHttpClient();
```

Comment:
Just wondering if the connection is going to happen for each processed request, if so do we need any further optimization?

Reply:
good call out. I've added a cache with a 10 minute ttl to avoid bumping into rate limits